### PR TITLE
fix(geometry): reject degenerate spatial sizes in depth_to_normals

### LINKED ISSUE #4398: `depth_to_normals` returns all-zero "unit normals" when `H = 1` or `W = 1`, and `nan` in float16
`depth_to_normals` documents a unit normal per pixel. When either spatial axis has length 1 the result is a zero vector in float32 and float64, and `nan` in float16 — silently, with the documented output shape.

## Repro

```python
import torch
from kornia.geometry.depth import depth_to_normals

K = torch.tensor([[[100.0, 0.0, 1.0], [0.0, 100.0, 0.0], [0.0, 0.0, 1.0]]])
for shape in [(1, 1, 1, 3), (1, 1, 3, 1), (1, 1, 1, 1), (1, 1, 2, 3)]:
    for dt in (torch.float32, torch.float16, torch.float64):
        n = depth_to_normals(torch.full(shape, 2.0, dtype=dt), K.to(dt))
        print(shape, str(dt)[6:], tuple(n.shape), n.norm(dim=1).flatten().float().tolist())
```

Executed on `main` `147ecc4e`, torch 2.14.0 (cpu; mps gives the same values):

| depth shape | float32 | float16 | float64 |
|---|---|---|---|
| `(1, 1, 1, 3)` (H = 1) | `(1, 3, 1, 3)`, norms `[0, 0, 0]` | `nan` everywhere | norms `[0, 0, 0]` |
| `(1, 1, 3, 1)` (W = 1) | `(1, 3, 3, 1)`, norms `[0, 0, 0]` | `nan` everywhere | norms `[0, 0, 0]` |
| `(1, 1, 1, 1)` | `(1, 3, 1, 1)`, norm `[0]` | `nan` | `[0]` |
| `(1, 1, 2, 3)` (control) | norms all `1.0` | `1.0` | `1.0` |

A flat plane at `H = 3` returns `[0, 0, 1]` at the centre, so the sign convention itself is fine; only the degenerate sizes are wrong.

## Mechanism

`spatial_gradient` along a length-1 axis is identically zero, so one of the two tangent vectors is zero, `torch.linalg.cross` returns zero, and `F.normalize(normals, dim=-1, p=2)` divides `0` by `max(‖0‖, 1e-12)`: `0` in float32/float64, and in float16 the `1e-12` floor underflows to `0`, giving `0 / 0 = nan`. Nothing checks `H >= 2` / `W >= 2`; `KORNIA_CHECK_SHAPE(depth, ["B", "1", "H", "W"])` accepts any positive size.

## Classification

**Broken output** — a zero "unit normal" (or `nan`) with the documented shape and no error. Idiom 4 (a degenerate input silently produces garbage instead of raising) and idiom 2 (the `1e-12` normalisation floor is not scaled to the working dtype).

## Disposition

n/a — single symbol. The focused fix is to reject `H < 2 or W < 2` with a `ShapeError` (a normal field needs two tangent directions); if a zero normal is the intended answer for the degenerate case, say so in the docstring and use a dtype-scaled floor so float16 returns zeros too. Either way the `test_convention_*` pin goes on the chosen contract.

Found in the batch-5 audit (row `5c-nm-05`, tracking issue #4262), recorded there as unfiled; the `W = 1` case originally reached this function through #4278's squeeze and now reaches it on its own after #4298.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by NaCr05**
## What does this pull request do?

`depth_to_normals` currently accepts singleton spatial axes even though a surface normal requires two tangent directions. These inputs can produce zero or non-finite normals; empty spatial dimensions fail later inside padding.

This implements the rejection policy suggested in the issue: raise `ShapeError` with the actual dimensions when `H < 2` or `W < 2`, before unprojection and spatial gradients. The normal computation for valid sizes is unchanged. The docstring and changelog describe the minimum dimensions, and the older #4278 pin keeps its projection/warping assertions while the new #4398 pin covers the normal-size contract.

## Related issue or discussion

Fixes #4398.

## What did you check?

Local environment: Windows, Python 3.14.6, PyTorch 2.14.0+cpu.

- Tests first: all 12 new invalid-size cases failed on the unmodified implementation; 12 valid-size controls passed.
- Full depth module, CPU float32/float64, including gradchecks and full-graph `torch.compile` tests using `aot_eager`: **304 passed, 2 xfailed**.
- Focused CPU float16 and bfloat16 checks with their known-failure profiles: **25 passed each**. These cover invalid dimensions, 2x2/2x3/3x2 planes, both depth meanings, shared/per-batch intrinsics, and the retained #4278 projection/warping pin. The profiles emit Linux/Windows platform compatibility warnings locally.
- Depth module docstring examples: **6 passed**.
- Full repository pre-commit checks: **passed**.
- Type check: **exit 0**, with one deprecation warning in the unchanged `kornia/feature/lightglue.py`.

```text
KORNIA_TEST_OPTIMIZER=aot_eager pixi run -e py314 test-module tests/geometry/test_depth.py --device=cpu --dtype=float32,float64 --runslow -q
pixi run -e py314 test-module tests/geometry/test_depth.py --device=cpu --dtype=float16 --xfail-known-failures --known-failure-profile=cpu-float16 -k 'require_two_spatial_axes or minimum_spatial_size_plane or keeps_the_w_one_axis_4278' -q
pixi run -e py314 test-module tests/geometry/test_depth.py --device=cpu --dtype=bfloat16 --xfail-known-failures --known-failure-profile=cpu-bfloat16 -k 'require_two_spatial_axes or minimum_spatial_size_plane or keeps_the_w_one_axis_4278' -q
pixi run -e py314 test-module --doctest-modules kornia/geometry/depth.py -q
pixi run -e py314 pre-commit-all
pixi run -e py314 typecheck --python .venv-py314
```

Local validation did not cover GPU execution, Inductor code generation, or the full repository test suite.

## How did AI help?

Codex assisted with investigation, implementation, tests, documentation, and the checks above. The new regression tests were run before the fix to verify the failure. Half-precision plane-test data were checked against both upstream main and the patch and adjusted for representable coordinates without relaxing the repository's default tolerances.

### CODE REVIEW

**Review by ducha-aiki — APPROVED** [review]
Reviewed exact head `685c90ba`.

The guard matches the disposition in #4398: `H < 2 or W < 2` raises `ShapeError` with the actual dimensions, placed right after the existing `KORNIA_CHECK_SHAPE` calls and before `depth_to_3d_v2`, so nothing is allocated first. `depth_to_3d`, `depth_to_3d_v2` and `warp_frame_depth` are correct at `W = 1` (the retained #4278 pin covers them) and are rightly out of scope. The valid path is byte-identical to main: 16 configurations (float32/float64, four asymmetric shapes, both `normalize_points`) give `torch.equal` outputs and gradients. `Fixes #4398` is right for a single-symbol issue; the documented-defect grep is empty in `kornia/`; the release note cites #4458 with old and new behaviour.

Compile: `dynamo.explain` on `depth_to_normals` reports one graph and zero breaks, and `fullgraph=True` works both static and `dynamic=True`; this PR adds the first dynamo test for the module, and all four new nodes pass under inductor.

Validation (isolated worktree, import origin verified): all 24 `test_convention_normals_require_two_spatial_axes_4398` nodes fail on `8fb9ad12` and pass here; the 24-node valid-size plane test is a value control with asymmetric K and H != W. `tests/geometry/test_depth.py` CPU float32/float64 with `KORNIA_TEST_OPTIMIZER=inductor` 304 passed, 2 xfailed; cpu-float16 and cpu-bfloat16 profiles 35 passed, 1 pre-existing xfail each; MPS float32 35 passed, 1 skipped; ruff clean. Not run: full suite, CUDA.

Unrelated and pre-existing, not for this PR: `ShapeError.__init__` is not dynamo-traceable, so under `fullgraph=True` any `ShapeError` raise (this guard and every `KORNIA_CHECK_SHAPE`) surfaces as `torch._dynamo.exc.Unsupported`. Identical on main.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Claude (Fable 5.1).
